services/orderService.js

```diff
@@ -0,0 +1,80 @@
+const { orderDao, cartDao, userDao } = require("../models")
+
+
+const getOrder = async ( userId ) => {
+    const [ cart ] = await cartDao.getCartToCheckProduct( userId )
+    if ( !cart ) {
+        const err = new Error("Shopping cart not in selected products")
+        err.statusCode = 403;
+        throw err
+    }
+
+    const user = await userDao.getOrderUserInfo( userId )
+    const products = await orderDao.getOrderProduct( userId )
+    const coupons = await orderDao.getCouponToUser( userId )
+
+    for ( const el of products ) {
+        el.price = el.price * el.quantity
+    }
+
+    user.product = products;
+    user.coupon = coupons;
+
+    return user
+}
+
+const getOrderList = async ( userId ) => {
+    return await orderDao.getOrderList( userId );
+}
+
+const createOrder = async ( userId, address, couponId, point, price ) => {
+    const products = await cartDao.getCartToCheckProduct( userId );
+    if ( products.length == 0 ) {
+        const err = new Error("Shopping cart not in selected products");
+        err.statusCode = 403;
+        throw err;
+    }
+
+    const user = await userDao.getOrderUserInfo( userId );
+    if ( user.point < point ) {
+        const err = new Error("There are few points")
+        err.statusCode = 400;
+        throw err;
+    }
+
+    const savePoint = price * 0.05 * user.grade
+    
+    if ( couponId ) {
+        const useCoupon = await userDao.getCoupon( userId, couponId );
+        if ( !useCoupon ) {
+            const err = new Error("Coupon does not exist");
+            err.statusCode = 400;
+            throw err;
+        }
+    }
+
+    const { insertId } = await orderDao.createOrderList( price, address );
+    await orderDao.createOrder( userId, products, insertId );
+    await cartDao.deleteCheckProduct( userId );
+    await userDao.updateCoupon( userId, couponId )
```

Comment:
transaction으로 처리해줘야 데이터가 안정적으로 보장될 수 있습니다.

아래 코드를 참고해서, createOrder 로직, deleteCheckProduct 로직, updateCoupon 로직이 하나의 transaction으로 처리될 수 있도록 만들어주세요~!
```javascript
// create a new query runner
const queryRunner = dataSource.createQueryRunner()

// establish real database connection using our new query runner
await queryRunner.connect()

// now we can execute any queries on a query runner, for example:
await queryRunner.query("SELECT * FROM users")

// we can also access entity manager that works with connection created by a query runner:
const users = await queryRunner.manager.find(User)

// lets now open a new transaction:
await queryRunner.startTransaction()

try {
    // execute some operations on this transaction:
    await queryRunner.manager.save(user1)
    await queryRunner.manager.save(user2)
    await queryRunner.manager.save(photos)

    // commit transaction now:
    await queryRunner.commitTransaction()
} catch (err) {
    // since we have errors let's rollback changes we made
    await queryRunner.rollbackTransaction()
} finally {
    // you need to release query runner which is manually created:
    await queryRunner.release()
}
```

Reply:
https://orkhan.gitbook.io/typeorm/docs/transactions

Reply:
트렌젝션 시도했으나.. 제대로 작동하지 않아서 보류 했습니다 주말 내로 다시 시도해 보겠습니다